### ISSUE TITLE
Potential `@fab/node-runtime` or `@fab/fetch` package

### ISSUE THREAD

**Issue by geelen** [fab-spec/fab#264]
I've been nipping around the edges of this but I think I just need to properly bite in and... chew... this delicious metaphor?

Prompted by this:

https://github.com/fab-spec/fab/pull/262/files/a6fa22dbd2a2f072175072eb29c35d969022fd8f#diff-544c6c6d49ee404d3d9fee9a9959c235

## Background

FAB uses the [`fetch` API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) from the browser to make server-side FAB code feel more like a magical server-side serviceworker than a Node app. It's extremely useful in breaking some long-held assumptions about server-side JS apps, _plus_ it makes sense given our V8 target runtime is incompatible with all the built-in Node packages, like `http`.

So far this implementation consists of the following global objects:

* [`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch)
* [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request)
* [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response)
* [`ReadableStream`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream)
* [`Headers`](https://developer.mozilla.org/en-US/docs/Web/API/Headers)

We also provide `URL`, `console`, `NODE_ENV`, `process.env.NODE_ENV`, `setTimeout`, `setImmediate`, `clearTimeout` as per https://fab.dev/kb/fab-runtime-environment#runtime-consideration

This is provided by `@fab/sandbox-node-vm` for running `@fab/server`, and repeated slightly differently in `@fab/deployer-aws-lambda`. It's also used internally in Linc, but then _not_ used in Cloudflare Workers environment (they have their own).

But ~our~ my approach so far has been AHAF (ad hoc as fuck). And it's causing some problems:

## Problems

This kind of code does not make me feel good: ([source](https://github.com/fab-spec/fab/blob/93f8825834f9b5089b0800410879625e9d20613b/packages/sandbox-node-vm/src/index.ts#L16-L27))

```ts
export function HybridReadableStream(...args: any[]) {
  const readableStream = new WebReadableStream(...args)
  return new Proxy(readableStream, {
    getPrototypeOf() {
      return Stream.Readable.prototype
    },
    get(target, prop, receiver) {
      if (prop === 'on') return () => {}
      return Reflect.get(target, prop, receiver)
    },
  })
}
```

The comment above it explains:

> We're using node-fetch under the hood, which has a hard-coded check on the body:
>   `response.body instanceof Stream`
> this means that our Web Streams ReadableStream gets .toString()-ed, which breaks.
> This lets FAB users use ReadableStream as if it was on the web, but this Node VM
> (which will be replaced by the V8::Isolate-based VM) can trick node-fetch into
> doing what we want. See https://github.com/node-fetch/node-fetch/pull/848

TL;DR, trying to mix and match `node-fetch` and an express server meant streaming responses weren't being passed through correctly. This fixed it.

Last week? This weekend? What is time? Anyway, recently I wrote this: ([source](https://github.com/fab-spec/fab/blob/e2e-testing-chunking/packages/server/src/index.ts#L90-L101))

```js
const response = fetch(...)
return Object.create(response, {
  body: {
    value: Object.create(response.body, {
      getReader: {
        get() {
          const webStream = nodeToWebStream(response.body)
          return webStream.getReader.bind(webStream)
        },
      },
    }),
  },
})
```

This is so that the following (perfectly valid) instance works inside a FAB:

```ts
// Fetch three URLs in parallel
const urls = ['/part.one', 'part.two', 'part.three']
const responses = await Promise.all(urls.map(url => fetch(url)))

// Stream them through to a response
const streams = responses.map(response => response.body)
return new ReadableStream({
  async start(controller) {
    for (const stream of streams) {
      if (!stream) {
        continue
      }
      const reader = stream.getReader()
      while (true) {
        const { done, value } = await reader.read()
        if (done) {
          break
        }
        if (value !== undefined) {
          controller.enqueue(value)
        }
      }
      reader.releaseLock()
    }
    controller.close()
  },
})
```

This is a paraphrased version of #241 by @yacinehmito. It depends upon the fact that `response.body` returns a `ReadableStream` and a FAB can return one of those so stitching them together is quite straightforward. More importantly _any code that purely uses fetch API global objects should be able to drop into a FAB without alteration_. Which would be great.

But, as it was, I needed to make `stream.getReader()` call `nodeToWebStream(response.body)` behind the scenes. (I couldn't do it by default because then `fetch(x).then(response => response.json())` would break, since that's `node-fetch` internals so it doesn't like you using a Web Stream there.

Anyway, it all just seems a bit too AHAF right now, so how do we fix it?

## `@fab/fetch` package?

This I think is the best solution: make a package that provides all the `fetch` APIs that is wholly self-consistent (i.e. any usage that would work in the browser/service worker/CF worker), just designed to run in a Node environment. [Unlike `node-fetch`](https://github.com/node-fetch/node-fetch#features), _implicit_ interop with Node should not be possible. I.e. not this:

> Use native Node streams for body, on both request and response.

This raises the question how to adapt it for the `@fab/server` express server. At the moment we have this code, which works because `node-fetch` returns Node streams:

```ts
if (body instanceof Stream) {
if (!response_headers['transfer-encoding'])
  res.set('transfer-encoding', 'chunked')

await new Promise((resolve, reject) => {
  body.on('data', (chunk) => res.write(chunk))
  body.on('error', reject)
  body.on('end', resolve)
})
res.end()
```

That'd need to be changed, although the code for that [is already there, I think](https://github.com/fab-spec/fab/blob/a6fa22dbd2a2f072175072eb29c35d969022fd8f/packages/server/src/index.ts#L171-L188).

This is the sort of thing I feel a bit trepidatious about. It's basically a mashup of https://github.com/whatwg/fetch and https://github.com/node-fetch/node-fetch, where it's designed to _run_ in a node environment but present a 100% browser-compatible facade. Much easier to write that maintain, but I guess it's just JS, how hard it can be?

## `@fab/node-runtime`

Rather than going for a low-level, spec-compatible `fetch` interface, we could instead... not do that. It might be smarter to make a higher-level "node-runtime" package that does whatever hacky dark magic it needs to to make things work, using `node-fetch` and Proxy objects and prototype hacking in order to make something look from the _outside_ as if it's a web Fetch object, but internally it's just `node-fetch` and Node streams.

This is basically extending the hacky business I've already been doing, but colocating it into one little package of ugliness. Hiding it behind a privacy screen, if you will.

On the plus side it means we can reuse a lot more code, from `node-fetch` of course but even things like https://github.com/xuset/readable-stream-node-to-web. On the minus, it sets us up for metaprogramming whack-a-mole in JS.

## Further considerations

One thing that's always been on the roadmap is to use a real `V8::Isolate` inside `@fab/server`, making it much more of a restricted, secure environment, and capable for multi-tenancy on a huge scale. The [current version](https://github.com/fab-spec/fab/blob/72e02d1ed251426bea67cbd115a33af0b716703a/packages/server/src/sandboxes/v8-isolate.ts) of that code is non-functional, but a lot of the fetch API groundwork has already been done by the (now deprecated) Fly project at https://github.com/superfly/flyjs/tree/master/packages/v8env

There's some surface-level differences, but the [core fetch logic](https://github.com/superfly/flyjs/tree/master/packages/core/src/bridge/fetch) is there and does work, I just haven't worked on this for a while. But potentially we don't need to continue to support Node at all, but _always_ interpret a FAB through a `V8::Isolate` container.

I'm not sure what this would mean for the upcoming `fab dev` environment (#188), or the concept of a standalone "FAB Server Spec" project. But... this issue is long enough for now. I'd welcome any suggestions on how to proceed! 🙏
